linera-ethereum/src/client.rs

```diff
@@ -1,112 +1,187 @@
 // Copyright (c) Zefchain Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
+use std::fmt::Debug;
+
 use alloy::{
     primitives::{Address, U256},
-    providers::{Provider, ProviderBuilder, RootProvider},
     rpc::types::eth::{
         request::{TransactionInput, TransactionRequest},
-        BlockId, BlockNumberOrTag, Filter,
+        BlockId, BlockNumberOrTag, Filter, Log,
     },
-    transports::http::reqwest::Client,
 };
-use alloy_primitives::Bytes;
-use url::Url;
+use alloy_primitives::{Bytes, U64};
+use async_trait::async_trait;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_json::value::RawValue;
 
 use crate::common::{event_name_from_expanded, parse_log, EthereumEvent, EthereumServiceError};
-pub type HttpProvider = RootProvider<alloy::transports::http::Http<Client>>;
 
-/// The Ethereum endpoint and its provider used for accessing the ethereum node.
-pub struct EthereumEndpoint<M> {
-    pub provider: M,
+/// A basic RPC client for making JSON queries
+#[async_trait]
+pub trait JsonRpcClient {
+    type Error: From<serde_json::Error>;
+
+    /// The inner function that has to be implemented and access the client
+    async fn request_inner(&self, payload: Vec<u8>) -> Result<Vec<u8>, Self::Error>;
+
+    /// The function doing the parsing of the input and output.
+    async fn request<T, R>(&self, method: &str, params: T) -> Result<R, Self::Error>
+    where
+        T: Debug + Serialize + Send + Sync,
+        R: DeserializeOwned + Send,
+    {
+        let payload = Request::new(method, params);
+        let payload = serde_json::to_vec(&payload)?;
+        let body = self.request_inner(payload).await?;
+        let result = serde_json::from_slice::<JsonRpcResponse>(&body)?;
+        let raw = result.result;
+        let res = serde_json::from_str(raw.get())?;
+        Ok(res)
+    }
 }
 
-// TODO(2011) We need to support via a trait.
-impl EthereumEndpoint<HttpProvider> {
-    /// Lists all the accounts of the Ethereum node.
-    pub async fn get_accounts(&self) -> Result<Vec<String>, EthereumServiceError> {
-        Ok(self
-            .provider
-            .get_accounts()
-            .await?
-            .into_iter()
-            .map(|x| format!("{:?}", x))
-            .collect::<Vec<_>>())
+#[derive(Serialize, Deserialize, Debug)]
+struct Request<'a, T> {
+    id: u64,
+    jsonrpc: &'a str,
+    method: &'a str,
+    params: T,
+}
+
+impl<'a, T> Request<'a, T> {
+    /// Creates a new JSON RPC request, the id does not matter
+    pub fn new(method: &'a str, params: T) -> Self {
```

Comment:
Should we actually be using the ID, and then verify that the response's ID matches?

Reply:
We can do that as well.